read-fonts/src/tables/gsub/closure.rs

```diff
@@ -241,247 +368,475 @@ impl GlyphClosure for LigatureSubstFormat1<'_> {
+impl GlyphClosure for ContextFormat1<'_> {
+    fn add_reachable_glyphs(&self, ctx: &mut ClosureCtx<'_>) -> Result<(), ReadError> {
         let coverage = self.coverage()?;
-        for seq in coverage
+        let Some(cur_glyphs) = intersect_coverage(&coverage, ctx.current_glyphs()) else {
+            return Ok(());
+        };
+
+        // now for each rule set that applies to a current glyph:
+        for (i, seq) in coverage
             .iter()
-            .zip(self.chained_seq_rule_sets().iter())
-            .filter_map(|(gid, seq)| seq.filter(|_| glyphs.contains(gid)))
+            .zip(self.rule_sets())
+            .enumerate()
+            .filter_map(|(i, (gid, seq))| {
+                seq.filter(|_| cur_glyphs.contains(gid)).map(|seq| (i, seq))
+            })
         {
-            for rule in seq?.chained_seq_rules().iter() {
+            for rule in seq?.rules() {
                 let rule = rule?;
-                if rule
-                    .input_sequence()
-                    .iter()
-                    .chain(rule.backtrack_sequence())
-                    .chain(rule.lookahead_sequence())
-                    .all(|gid| glyphs.contains(gid.get()))
-                {
-                    lookups.extend(
-                        rule.seq_lookup_records()
-                            .iter()
-                            .map(|rec| rec.lookup_list_index()),
-                    );
+                // skip rules if the whole input sequence isn't in our glyphset
+                if !rule.matches_glyphs(ctx.glyphs()) {
+                    continue;
+                }
+                // python calls this 'chaos'. Basically: if there are multiple
+                // lookups applied at a single position they can interact, and
+                // we can no longer trivially determine the state of the context
+                // at that point. In this case we give up, and assume that the
+                // second lookup is reachable by all glyphs.
+                let mut seen_sequence_indices = IntSet::new();
+                for lookup_record in rule.lookup_records() {
+                    let lookup_id = lookup_record.lookup_list_index();
+                    let sequence_idx = lookup_record.sequence_index();
+                    let active_glyphs = if !seen_sequence_indices.insert(sequence_idx) {
+                        // During processing, when we see an empty set we will replace
+                        // it with the full current glyph set
+                        None
+                    } else if sequence_idx == 0 {
+                        Some(IntSet::from([coverage.iter().nth(i).unwrap()]))
+                    } else {
+                        Some(IntSet::from([rule.input_sequence()
+                            [sequence_idx as usize - 1]
+                            .get()]))
+                    };
+                    ctx.add_todo(lookup_id, active_glyphs);
```

Comment:
After some thing, I would suggest doing recursion here instead of adding new todos and handle them in the final step, because a recursion lookup can keep growing the todos list, which could cause a timeout. By using recursion, we can add a hard limit on recursion depth. If &LookupList is added as a struct member, we should be able to just call glyph_closure on the child Lookup here. See also harfbuzz: https://github.com/harfbuzz/harfbuzz/blob/18ab0f552279159920d2cecbf97052e21fcb8c9a/src/hb-ot-layout-gsubgpos.hh#L92, https://github.com/harfbuzz/harfbuzz/blob/18ab0f552279159920d2cecbf97052e21fcb8c9a/src/hb-ot-layout-gsubgpos.hh#L174